clients/src/main/java/org/apache/kafka/common/requests/RequestUtils.java

```diff
@@ -90,4 +75,37 @@ public static boolean hasTransactionalRecords(ProduceRequest request) {
         }
         return new AbstractMap.SimpleEntry<>(hasIdempotentRecords, hasTransactionalRecords);
     }
+
+    public static MessageSizeAccumulator size(
+        ObjectSerializationCache serializationCache,
+        Message header,
+        short headerVersion,
+        Message apiMessage,
+        short apiVersion
+    ) {
+        MessageSizeAccumulator messageSize = new MessageSizeAccumulator();
+        if (header != null)
+            header.addSize(messageSize, serializationCache, headerVersion);
+        apiMessage.addSize(messageSize, serializationCache, apiVersion);
+        return messageSize;
+    }
+
+    public static ByteBuffer serialize(
+        Message header,
+        short headerVersion,
+        Message apiMessage,
+        short apiVersion
+    ) {
+        ObjectSerializationCache serializationCache = new ObjectSerializationCache();
+        MessageSizeAccumulator messageSize = RequestUtils.size(serializationCache, header, headerVersion, apiMessage, apiVersion);
+
+        ByteBuffer buffer = ByteBuffer.allocate(messageSize.totalSize());
```

Comment:
Should it subtract ```messageSize.zeroCopySize()```?

Reply:
Yeah, good point. I don't think it affects its current non test usage, but could result in a bug as the code changes. Will fix.

Reply:
Actually, I need to think about this more. Because `SendBuilder` is not being used, we actually need to copy all the bytes to the returned buffer. I need to check what happens for a message contains records.

Reply:
oh, you are right. It uses ```ByteBufferAccessor``` rather than ```SendBuilder```.  There is no zero copy.